### PR TITLE
CUDA IPC UCT [WIP] Don't review

### PR DESCRIPTION
**Pull request by Akshay-Venkatesh**
- added cuda-ipc uct
- gtest changes
- no cache (in follow-up PR)

@bureddy

### CODE REVIEW

**Comment by mellanox-github** [comment]
Can one of the admins verify this patch?

**Comment by yosefe** [comment]
ok to test

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/6901/ for details (Mellanox internal link).

**Comment by bureddy** [comment]
@Akshay-Venkatesh  can you check below failures

```
19:20:30 [  FAILED  ] tcp/uct_flush_test.put_bcopy_flush_iface_no_comp/0, where GetParam() = tcp/ens2f1
19:20:30 [  FAILED  ] tcp/uct_flush_test.am_zcopy_flush_ep_nb/0, where GetParam() = tcp/ens2f1
19:20:30 [  FAILED  ] tcp/uct_flush_test.am_pending_flush_nb/0, where GetParam() = tcp/ens2f1
```

```
19:07:26 [  FAILED  ] ud_mlx5/uct_fence_test.and64/0, where GetParam() = ud_mlx5/mlx5_1:1
19:07:26 [  FAILED  ] ud_mlx5/uct_fence_test.xor64/0, where GetParam() = ud_mlx5/mlx5_1:1
19:07:26 [  FAILED  ] ud_mlx5/uct_fence_test.add64/0, where GetParam() = ud_mlx5/mlx5_1:1
19:07:26 [  FAILED  ] ud_mlx5/uct_fence_test.or64/0, where GetParam() = ud_mlx5/mlx5_1:1
19:07:26 [  FAILED  ] ud_mlx5/uct_flush_test.am_flush_iface_no_comp/0, where GetParam() = ud_mlx5/mlx5_1:1
19:07:26 [  FAILED  ] ud_mlx5/uct_flush_test.am_pending_flush_nb/0, where GetParam() = ud_mlx5/mlx5_1:1
19:07:26 [  FAILED  ] ud_mlx5/uct_flush_test.am_zcopy_flush_ep_nb/0, where GetParam() = ud_mlx5/mlx5_1:1
19:07:26 [  FAILED  ] ud_mlx5/uct_flush_test.am_zcopy_flush_ep_no_comp/0, where GetParam() = ud_mlx5/mlx5_1:1
19:07:26 [  FAILED  ] ud_mlx5/uct_flush_test.put_bcopy_flush_iface_no_comp/0, where GetParam() = ud_mlx5/mlx5_1:1
19:07:26 [  FAILED  ] rc_mlx5/test_uct_peer_failure.two_pairs_send/0, where GetParam() = rc_mlx5/mlx5_1:1
19:07:26 [  FAILED  ] rc_mlx5/test_uct_peer_failure.peer_failure/0, where GetParam() = rc_mlx5/mlx5_1:1
19:07:26 [  FAILED  ] rc/test_tag.tag_cancel_noforce/0, where GetParam() = rc/mlx5_1:1
19:07:26 [  FAILED  ] rc/test_tag.tag_rndv_zcopy_tag_mask/0, where GetParam() = rc/mlx5_1:1
19:07:26 [  FAILED  ] rc/test_tag.tag_eager_zcopy_expected/0, where GetParam() = rc/mlx5_1:1
19:07:26 [  FAILED  ] rc/test_tag.sw_rndv_expected/0, where GetParam() = rc/mlx5_1:1
19:07:26 [  FAILED  ] rc/test_tag.tag_eager_bcopy_unexpected/0, where GetParam() = rc/mlx5_1:1
19:07:26 [  FAILED  ] rc/test_tag.tag_eager_bcopy_tag_mask/0, where GetParam() = rc/mlx5_1:1
19:07:26 [  FAILED  ] rc/test_tag.tag_send_no_tag/0, where GetParam() = rc/mlx5_1:1
19:07:26 [  FAILED  ] rc/test_tag.tag_eager_zcopy_wrong_tag/0, where GetParam() = rc/mlx5_1:1
19:07:26 [  FAILED  ] rc/test_tag.tag_eager_short_expected/0, where GetParam() = rc/mlx5_1:1
19:07:26 [  FAILED  ] rc/test_tag.sw_rndv_unexpected/0, where GetParam() = rc/mlx5_1:1
19:07:26 [  FAILED  ] rc/test_tag.tag_rndv_zcopy_unexpected/0, where GetParam() = rc/mlx5_1:1
19:07:26 [  FAILED  ] ud/uct_p2p_mix_test_alloc_methods.mix1000_odp/0, where GetParam() = ud/mlx5_1:1
19:07:26 [  FAILED  ] ud/uct_p2p_mix_test_alloc_methods.mix1000_rcache/0, where GetParam() = ud/mlx5_1:1
19:07:26 [  FAILED  ] rc/uct_p2p_mix_test_indirect_atomic.mix1000_indirect_atomic/0, where GetParam() = rc/mlx5_1:1
19:07:26 [  FAILED  ] ud_mlx5/uct_p2p_mix_test_indirect_atomic.mix1000_indirect_atomic/0, where GetParam() = ud_mlx5/mlx5_1:1

```
```

17:03:28 [ RUN      ] cuda_ipc/test_uct_stats.get_zcopy/0
17:03:28 [1524751408.933788] [vulcan-jenkins:40729:0]    cuda_ipc_ep.c:65   UCX  ERROR cuPointerGetAttribute((void *) &local_ptr_ctx, attrib, (CUdeviceptr) buffer) is failed. ret:invalid argument
17:03:28 /scrap/jenkins/workspace/hpc-ucx-pr-5/label/vulcan-jenkins/worker/1/contrib/../test/gtest/uct/test_stats.cc:118: Failure
17:03:28 Value of: UCS_INPROGRESS == status || UCS_OK == status
17:03:28   Actual: false
17:03:28 Expected: true
17:03:38 /scrap/jenkins/workspace/hpc-ucx-pr-5/label/vulcan-jenkins/worker/1/contrib/../test/gtest/uct/test_stats.cc:128: Failure
17:03:38 Value of: m_comp.count
17:03:38   Actual: 2
17:03:38 Expected: 1
```

```
6:34:49 In file included from /scrap/jenkins/workspace/hpc-ucx-pr-5/label/vulcan-jenkins/worker/3/contrib/../src/uct/cuda/cuda_ipc/cuda_ipc_iface.c:9:
16:34:49 /scrap/jenkins/workspace/hpc-ucx-pr-5/label/vulcan-jenkins/worker/3/contrib/../src/uct/cuda/cuda_ipc/cuda_ipc_ep.h:24:3: error: redefinition of typedef 'uct_cuda_ipc_rem_seg_t' is a C11 feature [-Werror,-Wtypedef-redefinition]
16:34:49 } uct_cuda_ipc_rem_seg_t;
16:34:49   ^
16:34:49 /scrap/jenkins/workspace/hpc-ucx-pr-5/label/vulcan-jenkins/worker/3/contrib/../src/uct/cuda/cuda_ipc/cuda_ipc_ep.h:15:38: note: previous definition is here
16:34:49 typedef struct uct_cuda_ipc_rem_seg  uct_cuda_ipc_rem_seg_t;
```

File: src/uct/cuda/cuda_ipc/cuda_ipc_ep.h [hunk 1]
```diff
@@ -1,35 +1,53 @@
...
+typedef struct uct_cuda_ipc_rem_seg  uct_cuda_ipc_rem_seg_t;
+
+
+typedef struct uct_cuda_ipc_rem_seg {
+    uct_cuda_ipc_rem_seg_t *next;
+    CUipcMemHandle         ph;         /* Memory handle of GPU memory */
+    CUdeviceptr            d_bptr;     /* Allocation base address */
+    size_t                 b_len;      /* Allocation size */
+    int                    dev_num;    /* GPU Device number */
+} uct_cuda_ipc_rem_seg_t;
```
Comment (bureddy) [inline src/uct/cuda/cuda_ipc/cuda_ipc_ep.h]: I think caching is not part of this PR. right?

Comment (Akshay-Venkatesh) [reply src/uct/cuda/cuda_ipc/cuda_ipc_ep.h]: Correct. I've missed this. I'll remove this.

**Comment by Akshay-Venkatesh** [comment]
@bureddy 
This is the configuration I'm using:
../configure --prefix=/home/akvenkatesh/ucx-github/build-cuda-ipc --enable-debug --enable-cma --enable-gtest --enable-logging --with-cuda=/usr/local/cuda --with-gdrcopy=/usr

Are the config flags correct?

Do I need to test this on a machine with two HCAs?

All the above-mentioned tests are passing on a haswell machine I'm trying.

$ ./test/gtest/gtest --gtest_filter=ud_mlx5/uct_flush_test* | tail -5
[----------] 10 tests from ud_mlx5/uct_flush_test (798 ms total)

[----------] Global test environment tear-down
[==========] 10 tests from 1 test case ran. (799 ms total)
[  PASSED  ] 10 tests.
$ ./test/gtest/gtest --gtest_filter=ud_mlx5/uct_fence_test* | tail -5
[----------] 8 tests from ud_mlx5/uct_fence_test (540 ms total)

[----------] Global test environment tear-down
[==========] 8 tests from 1 test case ran. (540 ms total)
[  PASSED  ] 8 tests.
$ ./test/gtest/gtest --gtest_filter=tcp/uct_flush_test* | tail -5
[----------] 20 tests from tcp/uct_flush_test (44 ms total)

[----------] Global test environment tear-down
[==========] 20 tests from 1 test case ran. (44 ms total)
[  PASSED  ] 20 tests.

$ ./test/gtest/gtest --gtest_filter=rc/test_tag* | tail -5
[----------] 21 tests from rc/test_tag (2009 ms total)

[----------] Global test environment tear-down
[==========] 21 tests from 1 test case ran. (2010 ms total)
[  PASSED  ] 21 tests.

$ ./test/gtest/gtest --gtest_filter=ud/uct_p2p_mix_test_alloc_methods.mix1000_odp/0 | tail -5
[----------] 1 test from ud/uct_p2p_mix_test_alloc_methods (90 ms total)

[----------] Global test environment tear-down
[==========] 1 test from 1 test case ran. (90 ms total)
[  PASSED  ] 1 test.
$ ./test/gtest/gtest --gtest_filter=ud/uct_p2p_mix_test_alloc_methods.mix1000_rcache/0 | tail -5
[----------] 1 test from ud/uct_p2p_mix_test_alloc_methods (92 ms total)

[----------] Global test environment tear-down
[==========] 1 test from 1 test case ran. (92 ms total)
[  PASSED  ] 1 test.
$ ./test/gtest/gtest --gtest_filter=rc/uct_p2p_mix_test_indirect_atomic.mix1000_indirect_atomic/0 | tail -5 
[----------] 1 test from rc/uct_p2p_mix_test_indirect_atomic (104 ms total)

[----------] Global test environment tear-down
[==========] 1 test from 1 test case ran. (105 ms total)
[  PASSED  ] 1 test.
$ ./test/gtest/gtest --gtest_filter=ud_mlx5/uct_p2p_mix_test_indirect_atomic.mix1000_indirect_atomic/0 | tail -5
[----------] 1 test from ud_mlx5/uct_p2p_mix_test_indirect_atomic (72 ms total)

[----------] Global test environment tear-down
[==========] 1 test from 1 test case ran. (72 ms total)
[  PASSED  ] 1 test.

$ ./test/gtest/gtest --gtest_filter=cuda_ipc/test_uct_stats* | tail -5
[     INFO ] Using random seed of 7282
Note: Google Test filter = cuda_ipc/test_uct_stats*
[==========] Running 0 tests from 0 test cases.
[==========] 0 tests from 0 test cases ran. (1 ms total)
[  PASSED  ] 0 tests.


$ ./test/gtest/gtest --gtest_filter=rc_mlx5/test_uct_peer_failure* | tail -5
[----------] 1 test from rc_mlx5/test_uct_peer_failure_multiple (19091 ms total)

[----------] Global test environment tear-down
[==========] 6 tests from 3 test cases ran. (28115 ms total)
[  PASSED  ] 6 tests.

But with this I see this in info:
[     INFO ] < Error on QP 0x1717 wqe[086]: Transport retry count exceeded (synd 0x15 vend 0x81) opcode SEND >
[     INFO ] < Error on QP 0x1713 wqe[086]: Transport retry count exceeded (synd 0x15 vend 0x81) opcode SEND >
[     INFO ] < Error on QP 0x170f wqe[086]: Transport retry count exceeded (synd 0x15 vend 0x81) opcode SEND >
[     INFO ] < Error on QP 0x170b wqe[086]: Transport retry count exceeded (synd 0x15 vend 0x81) opcode SEND >

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/6904/ for details (Mellanox internal link).

**Comment by bureddy** [comment]
@Akshay-Venkatesh 

```
22:26:36 [ RUN      ] cuda_ipc/test_uct_stats.get_zcopy/0
22:26:36 /scrap/jenkins/workspace/hpc-ucx-pr/label/vulcan-jenkins/worker/1/contrib/../test/gtest/uct/test_stats.cc:118: Failure
22:26:36 Value of: UCS_INPROGRESS == status || UCS_OK == status
22:26:36   Actual: false
22:26:36 Expected: true
22:26:46 /scrap/jenkins/workspace/hpc-ucx-pr/label/vulcan-jenkins/worker/1/contrib/../test/gtest/uct/test_stats.cc:128: Failure
22:26:46 Value of: m_comp.count
22:26:46   Actual: 2
22:26:46 Expected: 1
```

**Comment by Akshay-Venkatesh** [comment]
@bureddy 
I've made changes to test stats buffer allocation for cuda-ipc uct to be tested. Lmk if the change works

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/6905/ for details (Mellanox internal link).

File: src/uct/cuda/cuda_ipc/cuda_ipc_md.c [hunk 2]
```diff
@@ -82,36 +148,34 @@ static ucs_status_t uct_cuda_ipc_mem_dereg(uct_md_h md, uct_mem_h memh)
...
+        .mkey_pack         = uct_cuda_ipc_mkey_pack,
+        .mem_reg           = uct_cuda_ipc_mem_reg,
+        .mem_dereg         = uct_cuda_ipc_mem_dereg,
+        .is_mem_type_owned = uct_cuda_is_mem_type_owned,
     };
     static uct_md_t md = {
         .ops          = &md_ops,
         .component    = &uct_cuda_ipc_md_component
     };
-
```
Comment (bureddy) [inline src/uct/cuda/cuda_ipc/cuda_ipc_md.c]: keep empty line

File: src/uct/cuda/cuda_ipc/cuda_ipc_md.c [hunk 3]
```diff
@@ -82,36 +148,34 @@ static ucs_status_t uct_cuda_ipc_mem_dereg(uct_md_h md, uct_mem_h memh)
...
-    if ((cudaErr!= cudaSuccess) || (num_gpus == 0)) {
-        ucs_debug("Not found cuda devices");
+    status = UCT_CUDADRV_FUNC(cuDeviceGetCount(&num_gpus));
+    if ((UCS_OK != status) || (num_gpus == 0)) {
+        ucs_debug("not found cuda devices");
         *resources_p     = NULL;
         *num_resources_p = 0;
         return UCS_OK;
     }
-
```
Comment (bureddy) [inline src/uct/cuda/cuda_ipc/cuda_ipc_md.c]: keep empty line

File: src/uct/cuda/cuda_ipc/cuda_ipc_md.h [hunk 4]
```diff
@@ -1,29 +1,67 @@
...
  * See file LICENSE for terms.
  */
 
 #ifndef UCT_CUDA_IPC_MD_H
 #define UCT_CUDA_IPC_MD_H
 
 #include <uct/base/uct_md.h>
+#include <uct/cuda/base/cuda_md.h>
+#include <uct/cuda/base/cuda_iface.h>
+#include <cuda.h>
```
Comment (bureddy) [inline src/uct/cuda/cuda_ipc/cuda_ipc_md.h]: you can remove this. it is already included in cuda_iface.h

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/6910/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/6911/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/6912/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/6913/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/6914/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/6915/ for details (Mellanox internal link).
